src/visibility_tracker.rs

```diff
@@ -606,198 +595,204 @@ fn register_name<'a>(
-                .variants
-                .iter()
-                .chain(enum_.impls.iter())
-                .filter_map(|id| crate_.index.get(id))
-            {
-                visit_root_reachable_public_items(
-                    crate_,
-                    parents,
-                    traversal_state,
-                    currently_visited_items,
-                    inner,
-                    next_parent_id,
+            rustdoc_types::ItemEnum::Enum(enum_) => {
+                stack.reserve(enum_.variants.len());
+                stack.extend(
+                    enum_
+                        .variants
+                        .iter()
+                        .chain(enum_.impls.iter())
+                        .filter_map(|id| {
+                            crate_.index.get(id).map(|inner| Action::Process {
+                                item: inner,
+                                parent_id: next_parent_id,
+                            })
+                        }),
                 );
             }
-        }
-        rustdoc_types::ItemEnum::Union(union_) => {
-            for inner in union_
-                .fields
-                .iter()
-                .chain(union_.impls.iter())
-                .filter_map(|id| crate_.index.get(id))
-            {
-                visit_root_reachable_public_items(
-                    crate_,
-                    parents,
-                    traversal_state,
-                    currently_visited_items,
-                    inner,
-                    next_parent_id,
+            rustdoc_types::ItemEnum::Union(union_) => {
+                stack.reserve(union_.fields.len());
+                stack.extend(
+                    union_
+                        .fields
+                        .iter()
+                        .chain(union_.impls.iter())
+                        .filter_map(|id| {
+                            crate_.index.get(id).map(|inner| Action::Process {
+                                item: inner,
+                                parent_id: next_parent_id,
+                            })
+                        }),
                 );
             }
-        }
-        rustdoc_types::ItemEnum::Trait(trait_) => {
-            for inner in trait_.items.iter().filter_map(|id| crate_.index.get(id)) {
-                visit_root_reachable_public_items(
-                    crate_,
-                    parents,
-                    traversal_state,
-                    currently_visited_items,
-                    inner,
-                    next_parent_id,
-                );
+            rustdoc_types::ItemEnum::Trait(trait_) => {
+                stack.reserve(trait_.items.len());
+                stack.extend(trait_.items.iter().filter_map(|id| {
+                    crate_.index.get(id).map(|inner| Action::Process {
+                        item: inner,
+                        parent_id: next_parent_id,
+                    })
+                }));
```

Comment:
Do you think it'd be feasible to make some kind of refactor to deduplicate this `reserve() + extend() + crate_.index.get(id).map(Action::Process {...})` pattern?

This method was pretty complex to start with, and now includes even more complexity with a fair bit of extra repetition.

Reply:
Fixed (by removing the refactor)